storage/blockdevice/tests/TESTS/blockdevice/general_block_device/main.cpp

```diff
@@ -563,69 +563,75 @@ void test_erase_functionality()
     TEST_ASSERT(erase_value_int <= 255);
     uint8_t erase_value = (uint8_t)erase_value_int;
 
-    // Determine start_address
-    bd_addr_t start_address = sectors_addr[rand() % num_of_sectors];
-    utest_printf("start_address=0x%016" PRIx64 "\n", start_address);
-
-    // Determine data_buf_size
-    bd_size_t erase_size = block_device->get_erase_size(start_address);
-    TEST_ASSERT(erase_size > 0);
-    bd_size_t data_buf_size = erase_size;
-
-    // Allocate buffer for write test data
-    uint8_t *data_buf = new (std::nothrow) uint8_t[data_buf_size];
-    TEST_SKIP_UNLESS_MESSAGE(data_buf != NULL, "Not enough memory for test");
-
-    // Allocate buffer for read test data
-    uint8_t *out_data_buf = new (std::nothrow) uint8_t[data_buf_size];
-    TEST_SKIP_UNLESS_MESSAGE(out_data_buf != NULL, "Not enough memory for test");
+    // Use sectors at the the start, middle and end of the flash
+    bd_size_t bd_size = block_device->size();
+    TEST_ASSERT(bd_size > 0);
+    bd_addr_t unaligned_addresses[] { 0, bd_size / 2, bd_size - 1 };
+    for (bd_addr_t address : unaligned_addresses) {
+        // Determine erase size (sector size)
+        bd_size_t erase_size = block_device->get_erase_size(address);
+        TEST_ASSERT(erase_size > 0);
+
+        // Align sector address
+        address -= address % erase_size;
+        utest_printf("sector address=0x%016" PRIx64 ", erase_size=%" PRIu64 "B\n", address, erase_size);
+
+        // Allocate buffer for write test data
+        size_t data_buf_size = erase_size;
+        uint8_t *data_buf = new (std::nothrow) uint8_t[data_buf_size];
```

Comment:
Do we have `std::unique_ptr` available?

Reply:
Mbed OS doesn't seem to have written guidelines around the STL vs manual/C-style code, but it seems in practice uses of STL are relatively rare (but exists in a few places). Performance/overhead is one reason, but anything that may raise exceptions can't be used because of `-fno-exceptions`. If we do `make_unique()` to do allocation here, it can potentially throw `std::bad_alloc` which we can't catch and falls back to a crash. This prevents us from using a very large proportion of the STL.

It's a good question though. The existing code in SFDP.cpp uses `make_unique` and checks for `nullptr`, but this won't work so I should fix that.

Reply:
If we encounter an out of memory condition can we do anything else other than crash? I would've thought even trying to catch `std::bad_alloc` is extremely rare, as we can't really continue if we get a bad alloc anyway. Mbed OS in general seems to have a rather inconsistent usage of c++ vs c style code. I'd generally prefer to default to c++ when possible. 

I do see your point about the fact we have exceptions disabled so we lose any exception safety from `make_unique`. Could we not still use the RAII features of `unique_ptr` to guarantee this memory is freed instead of freeing it manually? We don't need to use the `make_unique` factory to create a `unique_ptr`, we can alloc the memory ourselves and the `unique_ptr` will ensure it's deleted. I don't see why `auto data_buf = std::unique_ptr<uint8_t[]>(new (std::nothrow) int[data_buf_size]);` wouldn't work, and then we still get to benefit from the automatic free.

Reply:
> If we encounter an out of memory condition can we do anything else other than crash?

Some dynamic allocations are not critical, so we should be able to catch them and return errors. Exceptions that are purely due to bugs in code (e.g. array out of bound) can be allowed to crash in my opinion.

> I don't see why auto data_buf = std::unique_ptr<uint8_t[]>(new (std::nothrow) int[data_buf_size]); wouldn't work, and then we still get to benefit from the automatic free.

Cool, I'll try that.

Reply:
> auto data_buf = std::unique_ptr<uint8_t[]>(new (std::nothrow) uint8_t[data_buf_size]);

This works! But this particular file does lots of dynamic allocations, so fixing them all would be refactoring, and changing this line only would make style inconsistent. I think we should create an issue for improving C++ coding in Mbed OS.